GeosCore/tracer_mod.F90

```diff
@@ -0,0 +1,760 @@
+       ELSE IF ( TRIM(SpcInfo%Src_Vert) == 'stratosphere' ) THEN
+
+          !$OMP PARALLEL DO                                                  &
+          !$OMP DEFAULT( SHARED                                             )&
+          !$OMP PRIVATE( I, J, L                                            )
+          DO L = 1, State_Grid%NZ
+          DO J = 1, State_Grid%NY
+          DO I = 1, State_Grid%NX
+
+             ! Set mask to zero outside of stratosphere
+             IF ( .not. State_Met%InStratosphere(I,J,L) ) &
+                Mask(I,J,L) = 0.0_fp
+
+          ENDDO
+          ENDDO
+          ENDDO
+          !$OMP END PARALLEL DO
+
+       ENDIF
+
+       !--------------------------------------------------------------------
+       ! Apply source
+       !--------------------------------------------------------------------
+       IF ( TRIM(SpcInfo%Src_Mode) == 'constant' ) THEN
+       
+          !$OMP PARALLEL DO                                                  &
+          !$OMP DEFAULT( SHARED                                             )&
+          !$OMP PRIVATE( I, J, L                                            )
+          DO L = 1, State_Grid%NZ
+          DO J = 1, State_Grid%NY
+          DO I = 1, State_Grid%NX
+
+             IF ( SpcInfo%Src_Add ) THEN
+
+                ! Add source
+                State_Chm%Species(N)%Conc(I,J,L) =      &
+                   State_Chm%Species(N)%Conc(I,J,L) +   &
+                   ( SpcInfo%Src_Value * Mask(I,J,L) )
+
+             ELSE
+
+                ! Replace value
+                State_Chm%Species(N)%Conc(I,J,L) =      &
+                   ( SpcInfo%Src_Value * Mask(I,J,L) )
+
+             ENDIF
+
+          ENDDO
+          ENDDO
+          ENDDO
+          !$OMP END PARALLEL DO
+
+       ELSE IF ( TRIM(SpcInfo%Src_Mode) == 'model_field' ) THEN
+
+          ! Special handling for stOX tracer
+          IF ( TRIM(SpcInfo%Name) == 'stOX' ) THEN
+
+             !$OMP PARALLEL DO                                               &
+             !$OMP DEFAULT( SHARED                                          )&
+             !$OMP PRIVATE( I, J, L                                         )
+             DO L = 1, State_Grid%NZ
+             DO J = 1, State_Grid%NY
+             DO I = 1, State_Grid%NX
+                ! Replace value
+                State_Chm%Species(N)%Conc(I,J,L) = O3(I,J,L) * Mask(I,J,L)
+             ENDDO
+             ENDDO
+             ENDDO
+             !$OMP END PARALLEL DO
+          ENDIF
+
+       ELSE IF ( TRIM(SpcInfo%Src_Mode) == 'maintain_mixing_ratio' ) THEN
+
+          ! To distrubute tracer uniformly on the surface, compute the
+          ! total area
+          DO J = 1, State_Grid%NY
+          DO I = 1, State_Grid%NX
+
+             ! Accumulate the total area [m2]
+             Local_Tally = Local_Tally + &
+                           ( State_Grid%Area_M2(I,J) * Mask(I,J,1) )
+
+          ENDDO
+          ENDDO
+
+#ifdef MODEL_GEOS
```

Comment:
This will need to be done for GCHP too should be #if defined( MODEL_GCHP ) || defined( MODEL_GEOS ).

Reply:
Now resolved in commit c32a6bba.